### PR TITLE
WIP Mailjet backend

### PR DESCRIPTION
**Pull request by calvin**
Still very early in the stage, but I opened this PR to get some feedbacks.

Todos:

- [ ] add mock server
- [ ] add tests

### CODE REVIEW

**Comment by medmunds** [comment]
Hey, this is looking great. Thanks for putting it together!

For backend tests, I usually just copy and modify the tests from a similar ESP. It looks like Mailjet's send API uses a straightforward JSON payload, so the Postmark tests could be a good starting point.

File: anymail/backends/mailjet.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+    def set_from_email(self, email):
+        self.data["FromEmail"] = email.email
+        if email.name:
+            self.data["FromName"] = email.name
+
+    def add_recipient(self, recipient_type, email):
+        assert recipient_type in ["to", "cc", "bcc"]
+        if email.name:
+            address = "{} <{}>".format(email.name, email.email)
```
Comment (medmunds) [inline anymail/backends/mailjet.py]: This *might* have character-escaping issues (if, say, name contains angle brackets or commas).

If the goal is a standard RFC-formatted name/email pair, you can just use `email.address` for both cases. It returns a properly formatted, quoted, and escaped recipient -- e.g., `"Edmunds, Mike" <medmunds@example.com>`. (Haven't checked Mailjet's docs to see what they expect.)

Comment (calvin) [inline anymail/backends/mailjet.py]: It expects one of the following formats:
 - "example@example.org"
 - "<example@example.org>"
 - "Doe, John <example@example.org>"

If `email.address` always gives the RFC-compatible name/email pair, I'll just use `email.address`!

**Comment by Lekensteyn** [comment]
FYI: we are evaluating MJ and try to use it with django-anymail. The addition is patched into a "working" state at https://github.com/CodeYellowBV/django-anymail/tree/feature_mailjet

"working" because it can solicit a 200 OK response from MJ, but I could not verify the arrival yet (due to incomplete SPF records).

@calvin Are you still working on this feature?

**Comment by calvin** [comment]
@Lekensteyn 

I'd like to add docs and tests but I'm extremely busy right now.

BTW, I saw your changes and they look very good. Did you receive any mails yet?

**Comment by Lekensteyn** [comment]
@calvin I might also give a shot at adding docs/tests as time permits (but help is of course always welcome :-))

And yes, I managed to receive mail, it was not delivered due to some non-intuitive details of MJ:

 - Subaccounts need a special permission grant from the master account.
 - Even if you can see sub-account keys, you cannot control the Sender addresses/domains.
 - Since it was not obvious that the sub-account key is completely separate from the main one, and the default allowed senders list is empty, mails were not sent. Since the current error reporting is lacking, mails just end up in :blackhole: with no feedback.

In a MJ support ticket, they mentioned that a 3.1 API version is coming which should improve error reporting (public beta for SendAPI 3.1 is estimated for May).

Another issue we noticed is that it is not possible to leave out the FromEmail address (in an attempt to use the address from the template). Doing so will cause an error (400 FromEmail or Sender are required), this will also be addressed in the next API version. Until this new API is ready, you would have to manually pull out the mail from the template.

This seems to be generally useful though, @medmunds is it OK to do another GET request to query the address for a given template which can then be filled into the parameters? When 3.1 is actually released, this hack can then be removed.

**Comment by Lekensteyn** [comment]
Added a quick GET request for the template in CodeYellowBV/django-anymail@2aa2b26d89bc3588d1777486ec96cd7792b07b02. At first I wanted to do a simple `requests.get`, but then figured it would be better to re-use the existing session to save a new TCP connection and TLS handshake. What do you think about this approach? Comments:
- A translation between requests and anymail exceptions is missing
- I considered using "post_to_esp" method with a newly crafted payload class, but that does not seem fitting here as we have no message to send, just a request.
- How can this be tested? Split everything until `json_response = ...` and then mock that?

As for supported Anymail features, do you have a template or list of features where details can be added? It should be pretty complete now (but there are some quirks due to Mailjet limitations).

What do you suggest for testing? Copy an existing file (which?) and adapt it to Mailjet?

**Comment by medmunds** [comment]
@Lekensteyn nice -- clever to reuse the requests session that's already set up. I think you made the right call on skipping post_to_esp.

Not sure what you mean by "translation between requests and anymail exceptions". The main point of using the Anymail exception classes is Anymail uses them to handle Django's weird fail_silently semantics. But you should get them for free since you're calling Anymail's raise_for_status and deserialize_json_response. [... oh, looking through it again, there's one more you'll need to catch -- I'll add a commit comment]

For tests, I'd suggest starting with [test_postmark_backend](https://github.com/anymail/django-anymail/blob/master/tests/test_postmark_backend.py) -- the Postmark send API is pretty similar (in shape) to Mailjet's, from what I can tell. 

Your template sender tests will probably run into RequestsBackendMockAPITestCase's [requests.Session.request patch](https://github.com/anymail/django-anymail/blob/master/tests/mock_requests_backend.py#L30-L42); I *think* you can set mock_request.side_effect to an iterable of return values for the series of request calls. (Unclear how mock [side_effect](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.side_effect) works if [return_value](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.return_value) is also set -- you might need to clear return_value.)

For the list of Anymail message features: any of the backend tests is a good list (either it tests that the feature works, or it tests that it raises AnymailUnsupportedFeature). Or look through the [BasePayload abstract implementation](https://github.com/anymail/django-anymail/blob/master/anymail/backends/base.py#L343-L438) -- anything your payload omits will be unsupported.

And if you have a list of quirks, that's incredibly helpful. (It's usually the hardest part of the ESP-specific docs.) We could handle the docs as a separate PR to avoid holding things up -- just keeping track of the details in this PR is a great start.

I signed up for a Mailjet account, so we can get live integration tests in Travis when you're ready.

Thanks for continuing to push on this.

**Comment by Lekensteyn** [comment]
If I'm not mistaken `self.backend.session.get` could throw a RequestException which needs to be converted into a Anymail exception (AnymailRequestsAPIError?), that is what I meant by translation.

I'll later have a look at adding tests and while doing that fixup the exception handling (test-driven development :)).

**Comment by medmunds** [comment]
Oh, of course -- you're absolutely right.

It would probably be sufficient to just raise an AnymailRequestsAPIError directly in that case, without going through the conversion from requests error like post_to_esp does. (The conversion is there for backwards compatibility with Djrill, where callers might have been looking for a specific requests error. I don't think that would apply to Mailjet users.)

**Comment by Lekensteyn** [comment]
Update: some more bugfixes, tests and fixups of the exception handling on https://github.com/CodeYellowBV/django-anymail/tree/feature_mailjet, here are changes that are not added to this PR yet:
8a64a10 Also handle different method of specifying Sender
fde35e7 mailjet: fix Reply-To support
4eb9227 mailjet: allow setting extra ESP fields
28ae51c mailjet: add metadata and change tags implementation
fc3222a tests: add initial Mailjet backend tests
f45a4b5 tests: add Mailjet test for template responses
0c4d418 Mailjet: convert errors from template API

Tested with: `./runtests.py tests.test_mailjet_backend`

Still needed is an integration test which is harder since there is no testing environment or "sandbox":
https://twitter.com/seanhagen/status/435954163235557377. Furthermore, the API responses isn't that great, it always reports "Sent" even though a mail got held (because the sender address is not whitelisted for example). So for checking the actual status, one would have to do additional API queries (in polling mode).

(Maybe I should open a separate PR such that Travis can automatically check the changes as I push changes?)

**Comment by medmunds** [comment]
@Lekensteyn this is looking amazing! And very complete.

Yes, I think it might be easier to open a new PR on your branch.

Here's how I've been using live integration tests with the other ESPs:

* Most ESPs *don't* offer a sandbox. So the tests run truly live. The recipients are always \@example.com (or SendGrid's special "sink domain" for their tests) to avoid test email arriving in unexpected places.
* I maintain an Anymail test account and validated sender domain for each ESP (that doesn't make this cost-prohibitive). The API keys for those accounts are available in the Travis environment on the master branch. (Travis security blocks those environment vars in most pull requests, so the live tests won't run from your PR until I merge it. But you can set the vars in your environment to your own account for local testing.)
* Most ESPs *don't* return any useful send status from their send APIs. (Email is inherently async, and you have to implement webhooks if you want to know actual send status.) So the live integration tests are mainly making sure the ESP API hasn't changed out from under Anymail (isn't rejecting Anymail payloads). Travis checks this weekly, and it has caught some changes.
* The other big value of the live tests is debugging the initial implementation. While I'm developing, I add myself as a recipient to the [test_all_options](https://github.com/anymail/django-anymail/blob/master/tests/test_sendgrid_integration.py#L61-L89) live test, and set the environment variables to my own ESP account -- and then carefully inspect the mail I receive and the ESP API logs to make sure everything matches the test message. (And running that test usually uncovers some misunderstanding I had about their API.) Same thing for templates and test_merge_data. Just don't forget to remove yourself before committing.

**Comment by calvin** [comment]
Hi guys,

The latest changes from https://github.com/CodeYellowBV/django-anymail/commits/feature_mailjet are merged and pushed to github.

**Comment by dhrp** [comment]
@medmunds, @calvin @Lekensteyn, is there anything left to do? I've successfully implemented this branch in our project and didn't have any problems. Perhaps we can merge it?

I'd be happy to help, if necessary.

**Comment by Lekensteyn** [comment]
@dhrp Functionality-wise it should be ready. One part of the test suite is updated, but an integration test with Mailjet is still needed. It would be great if you could help on that front 😃

**Comment by medmunds** [comment]
@Lekensteyn Travis seems to be failing on some of your functional tests -- looks like mainly around templates. Not sure what might be different from your local environment.

(Sorry, I figured you'd seen those errors after @calvin merged your branch into this PR, but you probably don't get notified. Check the build status at the bottom of the PR on GitHub.)

**Comment by lorddaedra** [comment]
Please refactor to use official https://github.com/mailjet/mailjet-apiv3-python

**Comment by Lekensteyn** [comment]
@medmunds I'll have a look at the tests.

@lorddaedra We investigated that option before, but it is just a very thin wrapper around the REST API which duplicates some parts of anymail. What would be the benefit of using that wrapper?

**Comment by Lekensteyn** [comment]
@medmunds / @calvin Can you fetch the latest branch (with commit a15f1ec933e3c663fbcd15dc4b641740cc48769d) to fix the tests?

**Comment by medmunds** [comment]
@Lekensteyn could I suggest you open a new PR directly from your branch? (I think it'll be easier for you to have direct control. The git history will transfer, so everyone who contributed will still get credit.)

Also, if you haven't recently, it would probably be helpful to rebase your branch onto (this repo's) master before opening the PR.

**Comment by medmunds** [comment]
Superseded by #66.

**Pull request closed without merging**